File: src/main/infrastructure/persistence/query-store.ts
/**
 * JSON persistence for daily query records.
 *
 * Serialization format is intentionally compatible with the Python
 * implementation so both can read each other's files.
 */

import { readFile, writeFile, mkdir } from 'node:fs/promises';
import path from 'node:path';
import { getAppDataPath } from '../config.js';
import type { QueryStoragePort } from '../../domain/ports.js';
import {
  DailyQueryRecord,
  getDotNetMoniker,
  dotNetVersionFromMoniker,
} from '../../domain/models.js';
import type { DotNetVersion, CodeSearchHit } from '../../domain/models.js';

// ---------------------------------------------------------------------------
// Legacy moniker migration (files written before the ".0" suffix fix)
// ---------------------------------------------------------------------------

const LEGACY_MAP: Record<string, string> = {
  net5: 'net5.0',
  net6: 'net6.0',
  net7: 'net7.0',
  net8: 'net8.0',
  net9: 'net9.0',
  net10: 'net10.0',
};

function migrateMoniker(moniker: string): string {
  return LEGACY_MAP[moniker] ?? moniker;
}

// ---------------------------------------------------------------------------
// Date helpers — JSON stores dates as "YYYY-MM-DD"
// ---------------------------------------------------------------------------

function formatDate(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

function parseDate(iso: string): Date {
  const [y, m, d] = iso.split('-').map(Number) as [number, number, number];
  return new Date(y, m - 1, d);
}

// ---------------------------------------------------------------------------
// Serialization helpers (snake_case JSON ↔ camelCase TS)
// ---------------------------------------------------------------------------

interface HitDict {
  repository_name: string;
  project_name: string;
  dotnet_version: string;
  branch: string;
  csproj_count?: number;
}

interface RecordDict {
  query_date: string;
  organization: string;
  project: string | null;
  versions_searched: string[];
  results_by_version: Record<string, HitDict[]>;
}

function hitToDict(hit: CodeSearchHit): HitDict {
  return {
    repository_name: hit.repositoryName,
    project_name: hit.projectName,
    dotnet_version: getDotNetMoniker(hit.dotnetVersion),
    branch: hit.branch,
    csproj_count: hit.csprojCount,
  };
}

function hitFromDict(d: HitDict): CodeSearchHit {
  const moniker = migrateMoniker(d.dotnet_version ?? 'net8.0');
  const version = dotNetVersionFromMoniker(moniker);
  return {
    repositoryName: d.repository_name ?? '',
    projectName: d.project_name ?? '',
    dotnetVersion: version ?? ('NET_8' as DotNetVersion),
    branch: d.branch ?? '',
    csprojCount: d.csproj_count ?? 1,
  };
}

function recordToDict(record: DailyQueryRecord): RecordDict {
  const resultsByVersion: Record<string, HitDict[]> = {};
  for (const [key, hits] of Object.entries(record.resultsByVersion)) {
    resultsByVersion[key] = hits.map(hitToDict);
  }
  return {
    query_date: formatDate(record.queryDate),
    organization: record.organization,
    project: record.project,
    versions_searched: record.versionsSearched.map((v) => getDotNetMoniker(v)),
    results_by_version: resultsByVersion,
  };
}

function recordFromDict(d: RecordDict): DailyQueryRecord {
  // Migrate versions_searched
  const versionsSearched: DotNetVersion[] = [];
  for (const raw of d.versions_searched ?? []) {
    const migrated = migrateMoniker(raw);
    const v = dotNetVersionFromMoniker(migrated);
    if (v !== undefined) {
      versionsSearched.push(v);
    }
  }

  // Migrate results_by_version keys and parse hits
  const resultsByVersion: Record<string, CodeSearchHit[]> = {};
  for (const [key, hits] of Object.entries(d.results_by_version ?? {})) {
    const migratedKey = migrateMoniker(key);
    const parsed = hits.map(hitFromDict);
    if (migratedKey in resultsByVersion) {
      resultsByVersion[migratedKey]!.push(...parsed);
    } else {
      resultsByVersion[migratedKey] = parsed;
    }
  }

  return new DailyQueryRecord({
    queryDate: parseDate(d.query_date),
    organization: d.organization ?? '',
    project: d.project ?? null,
    versionsSearched,
    resultsByVersion,
  });
}

// ---------------------------------------------------------------------------
// JsonQueryStore
// ---------------------------------------------------------------------------

export class JsonQueryStore implements QueryStoragePort {
  private readonly filePath: string;

  constructor() {
    this.filePath = path.join(getAppDataPath(), 'queries.json');
  }

  async saveQuery(record: DailyQueryRecord): Promise<void> {
    const records = await this.loadQueries();
    const dateStr = formatDate(record.queryDate);
    const existing = records.find((r) => formatDate(r.queryDate) === dateStr);

    if (existing) {
      existing.merge(record);
    } else {
      records.push(record);
    }

    await this.write(records);
  }

  async loadQueries(): Promise<DailyQueryRecord[]> {
    try {
      const raw = await readFile(this.filePath, 'utf-8');
      if (!raw.trim()) return [];
      const data = JSON.parse(raw) as RecordDict[];
      return data.map(recordFromDict);
    } catch {
      return [];
    }
  }

  async getQueryByDate(queryDate: Date): Promise<DailyQueryRecord | null> {
    const records = await this.loadQueries();
    const target = formatDate(queryDate);
    return records.find((r) => formatDate(r.queryDate) === target) ?? null;
  }

  private async write(records: DailyQueryRecord[]): Promise<void> {
    await mkdir(path.dirname(this.filePath), { recursive: true });
    const data = records.map(recordToDict);
    await writeFile(this.filePath, JSON.stringify(data, null, 2), 'utf-8');
  }
}

File: src/azure_repos_scan/infrastructure/persistence/query_store.py
"""Persistencia de consultas históricas en JSON local."""

from __future__ import annotations

import json
from datetime import date
from pathlib import Path
from typing import Any, final

from azure_repos_scan.domain.models import (
    CodeSearchHit,
    DailyQueryRecord,
    DotNetVersion,
)

_VERSION_MAP: dict[str, DotNetVersion] = {v.value[1]: v for v in DotNetVersion}

# Monikers legacy (antes del fix .0) → moniker actual
_LEGACY_MAP: dict[str, str] = {
    "net5": "net5.0",
    "net6": "net6.0",
    "net7": "net7.0",
    "net8": "net8.0",
    "net9": "net9.0",
    "net10": "net10.0",
}


def _get_default_path() -> Path:
    """Ruta por defecto: %APPDATA%/azure-repos-scan/queries.json."""
    import os

    appdata = os.environ.get("APPDATA", str(Path.home()))
    folder = Path(appdata) / "azure-repos-scan"
    folder.mkdir(parents=True, exist_ok=True)
    return folder / "queries.json"


def _hit_to_dict(hit: CodeSearchHit) -> dict[str, str]:
    return {
        "repository_name": hit.repository_name,
        "project_name": hit.project_name,
        "dotnet_version": hit.dotnet_version.moniker,
        "branch": hit.branch,
    }


def _hit_from_dict(d: dict[str, Any]) -> CodeSearchHit:
    moniker: str = d.get("dotnet_version", "net8.0")
    moniker = _LEGACY_MAP.get(moniker, moniker)
    version = _VERSION_MAP.get(moniker, DotNetVersion.NET_8)
    return CodeSearchHit(
        repository_name=d.get("repository_name", ""),
        project_name=d.get("project_name", ""),
        dotnet_version=version,
        branch=d.get("branch", ""),
    )


def _record_to_dict(record: DailyQueryRecord) -> dict[str, Any]:
    return {
        "query_date": record.query_date.isoformat(),
        "organization": record.organization,
        "project": record.project,
        "versions_searched": [v.moniker for v in record.versions_searched],
        "results_by_version": {
            key: [_hit_to_dict(h) for h in hits]
            for key, hits in record.results_by_version.items()
        },
    }


def _record_from_dict(d: dict[str, Any]) -> DailyQueryRecord:
    versions_raw: list[str] = d.get("versions_searched", [])
    versions = [
        _VERSION_MAP[_LEGACY_MAP.get(m, m)]
        for m in versions_raw
        if _LEGACY_MAP.get(m, m) in _VERSION_MAP
    ]
    results_raw: dict[str, list[dict[str, Any]]] = d.get("results_by_version", {})
    results: dict[str, list[CodeSearchHit]] = {}
    for key, hits in results_raw.items():
        migrated_key = _LEGACY_MAP.get(key, key)
        parsed = [_hit_from_dict(h) for h in hits]
        if migrated_key in results:
            results[migrated_key].extend(parsed)
        else:
            results[migrated_key] = parsed
    return DailyQueryRecord(
        query_date=date.fromisoformat(d["query_date"]),
        organization=d.get("organization", ""),
        project=d.get("project"),
        versions_searched=versions,
        results_by_version=results,
    )


@final
class JsonQueryStore:
    """Almacena consultas históricas en un archivo JSON local."""

    def __init__(self, path: Path | None = None) -> None:
        self._path = path or _get_default_path()

    def save_query(self, record: DailyQueryRecord) -> None:
        """Guardar o merge consulta del día."""
        records = self.load_queries()
        existing = next(
            (r for r in records if r.query_date == record.query_date),
            None,
        )
        if existing is not None:
            existing.merge(record)
        else:
            records.append(record)
        self._write(records)

    def load_queries(self) -> list[DailyQueryRecord]:
        """Cargar todas las consultas históricas."""
        if not self._path.exists():
            return []
        raw_text = self._path.read_text(encoding="utf-8")
        if not raw_text.strip():
            return []
        data: list[dict[str, Any]] = json.loads(raw_text)
        return [_record_from_dict(d) for d in data]

    def get_query_by_date(self, query_date: date) -> DailyQueryRecord | None:
        """Obtener consulta de un día específico."""
        records = self.load_queries()
        return next((r for r in records if r.query_date == query_date), None)

    def _write(self, records: list[DailyQueryRecord]) -> None:
        """Escribir todas las consultas al archivo JSON."""
        self._path.parent.mkdir(parents=True, exist_ok=True)
        data = [_record_to_dict(r) for r in records]
        self._path.write_text(
            json.dumps(data, ensure_ascii=False, indent=2),
            encoding="utf-8",
        )
